File: examples/Simple/client/cpp/BasicStatic/SimpleClient.cpp
#include <thrift/protocol/TBinaryProtocol.h>
#include <thrift/transport/TSocket.h>
#include <thrift/transport/TTransportUtils.h>
#include <stdlib.h>

#include "../gen-cpp/SimpleService.h"

using namespace std;
using namespace apache::thrift;
using namespace apache::thrift::protocol;
using namespace apache::thrift::transport;

using namespace ::com::maxeler::Simple;

int check(std::vector<double> dataOutDFE, std::vector<double> dataOutCPU, int size) {
	int status = 0;

	for(int i=0; i < size; i++) {
		if(dataOutDFE[i] != dataOutCPU[i]) {
			fprintf(stderr, "Output data @ %d = %1.8g (expected %1.8g)\n",
				i, dataOutDFE[i], dataOutCPU[i]);
			status = 1;
		}
	}

	return status;
}

std::vector<double> SimpleCPU(int size, std::vector<double> dataIn) {
	std::vector<double> dataOut;
	dataOut.resize(size);

	for (int i=0 ; i<size ; i++) {
		dataOut[i] = dataIn[i]*dataIn[i] + dataIn[i];
	}

	return dataOut;
}

std::vector<double> SimpleDFE(int size, std::vector<double> dataIn) {
	// Make socket
	boost::shared_ptr<TTransport> socket(new TSocket("localhost", 9090));

	// Buffering is critical. Raw sockets are very slow
	boost::shared_ptr<TTransport> transport(new TBufferedTransport(socket));

	// Wrap in a protocol
	boost::shared_ptr<TProtocol> protocol(new TBinaryProtocol(transport));

	// Create a client to use the protocol encoder
	SimpleServiceClient client(protocol);

	std::vector<double> dataOut;
	dataOut.resize(size);

	try {
		// Connect!
		transport->open();

		// Allocate and send input streams to server
		remote_ptr address_dataIn = client.malloc_float(size);
		client.send_data_float(address_dataIn, dataIn);

		// Allocate memory for output stream on server
		remote_ptr address_dataOut = client.malloc_float(size);

		cout << "Running DFE.\n";

		client.Simple(size, address_dataIn, address_dataOut);

		// Get output stream from server
	        client.receive_data_float(dataOut, address_dataOut, size);

		// Free allocated memory for streams on server
		client.free(address_dataIn);
        	client.free(address_dataOut);

		// Free allocated maxfile data
		client.Simple_free();
	
		// Close!
		transport->close();
    
	} catch (TException& thrift_exceptiion) {
		cout << "ERROR: " << thrift_exceptiion.what() << endl;
		exit(-1);
	}

	return dataOut;
}

int main() {
	// Input
	const int size = 1024;
	std::vector<double> dataIn;

	for(int i = 0; i < size; i++) {
		dataIn.push_back(i + 1);
	}

	// CPU Output
	std::vector<double> dataOutCPU = SimpleCPU(size, dataIn);

	// DFE Output
        std::vector<double> dataOutDFE = SimpleDFE(size, dataIn);

	// Checking results
	if (check(dataOutDFE, dataOutCPU, size)) {
		cout << "Test failed." << endl;
		exit(-1);
	} else {
		cout << "Test passed!" << endl;
	}

	return 0;
}

File: examples/Simple/client/php/BasicStatic/SimpleClient.php
<?php 
set_include_path(get_include_path() . PATH_SEPARATOR . getenv('PHP_THRIFT_LIB'));

require_once 'Thrift/ClassLoader/ThriftClassLoader.php';

use Thrift\ClassLoader\ThriftClassLoader;

$GEN_DIR = '../gen-php';

$loader = new ThriftClassLoader();
$loader->registerNamespace('Thrift', getenv('PHP_THRIFT_LIB'));
$loader->registerDefinition('com\maxeler\Simple', $GEN_DIR);
$loader->register();

use Thrift\Protocol\TBinaryProtocol;
use Thrift\Transport\TSocket;
use Thrift\Transport\TBufferedTransport;
use Thrift\Exception\TException;

use com\maxeler\Simple\SimpleServiceClient;

function check ($dataOutDFE, $dataOutCPU, $size) {
    $status = 0;

    for($i=0; $i < $size; $i++) {
        if($dataOutDFE[$i] != $dataOutCPU[$i]) {
            echo "Output data @ ".$i." = ".$dataOutDFE[$i]." (expected ".$dataOutCPU[$i].")\r\n";
            $status++;
        }
    }
        return $status;

}

function SimpleCPU ($size, $dataIn) {
    for ($i = 0 ; $i < $size ; $i++) {
        $dataOut[$i] = $dataIn[$i] * $dataIn[$i] + $dataIn[$i];
    }

    return $dataOut;
}

function SimpleDFE ($size, $dataIn) {
    try {
	$startTime = microtime(true);
	
	# Make socket
        $socket = new TSocket('localhost', 9090);

	# Buffering is critical. Raw sockets are very slow
        $transport = new TBufferedTransport($socket);

	# Wrap in a protocol
        $protocol = new TBinaryProtocol($transport);

	# Create a client to use the protocol encoder
        $client = new SimpleServiceClient($protocol);

	echo "Creating a client:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";
	
	# Connect!
        $startTime = microtime(true);
        $transport->open();
	echo "Opening connection:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

        # Allocate and send input streams to server
	$startTime = microtime(true);
        $address_dataIn = $client->malloc_float($size);
        $client->send_data_float($address_dataIn, $dataIn);
	echo "Sending input data:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

        # Allocate memory for output stream on server
	$startTime = microtime(true);
        $address_dataOut = $client->malloc_float($size);
	echo "Allocating memory for output stream on server:\t".(microtime(true) - $startTime)."s\r\n";
	
	# Action default
        $startTime = microtime(true);
        $client->Simple($size, $address_dataIn, $address_dataOut);
	echo "Simple time:\t\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

        # Get output stream from server
	$startTime = microtime(true);
        $dataOut = $client->receive_data_float($address_dataOut, $size);
	echo "Getting output stream:\t(size = ".($size * 32)." bit)\t".(microtime(true) - $startTime)."s\r\n";

        # Free allocated memory for streams on server
	$startTime = microtime(true);
        $client->free($address_dataIn);
        $client->free($address_dataOut);
	echo "Freeing allocated memory for streams on server:\t".(microtime(true) - $startTime)."s\r\n";

	$startTime = microtime(true);
        $transport->close();
	echo "Closing connection:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

    } catch (TException $tx) {
        echo "ThriftException: ".$tx->getMessage()."\r\n";
        exit(-1);
    }

        return $dataOut;
}

$startTime = microtime(true);
$size = 1024;

for($i = 0; $i < $size; $i++) {
    $dataIn[$i] = $i + 1;
}
echo "Generating input data:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

# DFE Output
$startTime = microtime(true);
$dataOutDFE = SimpleDFE($size, $dataIn);
echo "DFE pass through total time:\t\t\t".(microtime(true) - $startTime)."s\r\n";

# CPU Output
$startTime = microtime(true);
$dataOutCPU = SimpleCPU($size, $dataIn);
echo "CPU pass through total time:\t\t\t".(microtime(true) - $startTime)."s\r\n";

# Checking results
$startTime = microtime(true);
$status = check($dataOutDFE, $dataOutCPU, $size);
echo "Checking results:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

if ($status == 0) { 
    echo "Test successful!\r\n";
} else {
    echo "Test failed ", $status, " times!\r\n";
    exit (-1);
}
